File: src/csharp/Intel/Iced.UnitTests/Intel/InstructionInfoTests/MemorySizeInfoTestReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using Iced.Intel;

namespace Iced.UnitTests.Intel.InstructionInfoTests {
	static class MemorySizeInfoTestReader {
		static readonly char[] commaSeparator = new char[] { ',' };
		static readonly char[] spaceSeparator = new char[] { ' ' };

		public static MemorySizeInfoTestCase[] GetTestCases() {
			var result = new MemorySizeInfoTestCase[IcedConstants.MemorySizeEnumCount];

			var filename = PathUtils.GetTestTextFilename("MemorySizeInfo.txt", "InstructionInfo");
			Debug.Assert(File.Exists(filename));
			int lineNo = 0;
			foreach (var line in File.ReadLines(filename)) {
				lineNo++;
				if (line.Length == 0 || line[0] == '#')
					continue;

				MemorySizeInfoTestCase tc;
				try {
					tc = ParseLine(line, lineNo);
				}
				catch (Exception ex) {
					throw new Exception($"Invalid line {lineNo} ({filename}): {ex.Message}");
				}
				if (result[(int)tc.MemorySize] is not null)
					throw new InvalidOperationException($"Duplicate test, {filename}, line {lineNo}");
				result[(int)tc.MemorySize] = tc;
			}

			var sb = new StringBuilder();
			for (int i = 0; i < result.Length; i++) {
				var tc = result[i];
				if (tc is null) {
					if (sb.Length > 0)
						sb.Append(", ");
					sb.Append(((MemorySize)i).ToString());
				}
			}
			if (sb.Length != 0)
				throw new InvalidOperationException($"Missing tests in {filename}: " + sb.ToString());
			return result!;
		}

		static MemorySizeInfoTestCase ParseLine(string line, int lineNo) {
			Static.Assert(MiscInstrInfoTestConstants.MemorySizeElemsPerLine == 6 ? 0 : -1);
			var elems = line.Split(commaSeparator, MiscInstrInfoTestConstants.MemorySizeElemsPerLine);
			if (elems.Length != MiscInstrInfoTestConstants.MemorySizeElemsPerLine)
				throw new Exception($"Expected {MiscInstrInfoTestConstants.MemorySizeElemsPerLine - 1} commas");

			var tc = new MemorySizeInfoTestCase();
			tc.LineNumber = lineNo;
			tc.MemorySize = ToEnumConverter.GetMemorySize(elems[0].Trim());
			tc.Size = NumberConverter.ToInt32(elems[1].Trim());
			tc.ElementSize = NumberConverter.ToInt32(elems[2].Trim());
			tc.ElementType = ToEnumConverter.GetMemorySize(elems[3].Trim());
			tc.ElementCount = NumberConverter.ToInt32(elems[4].Trim());
			foreach (var value in elems[5].Split(spaceSeparator, StringSplitOptions.RemoveEmptyEntries)) {
				if (!InstructionInfoDicts.MemorySizeFlagsTable.TryGetValue(value, out var flags))
					throw new InvalidOperationException($"Invalid flags value: {value}");
				tc.Flags |= flags;
			}
			return tc;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/info/MemorySizeInfoTestReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

import java.util.ArrayList;

import com.github.icedland.iced.x86.FileUtils;
import com.github.icedland.iced.x86.NumberConverter;
import com.github.icedland.iced.x86.PathUtils;
import com.github.icedland.iced.x86.ToMemorySize;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class MemorySizeInfoTestReader {
	public static MemorySizeInfoTestCase[] getTestCases() {
		String filename = PathUtils.getTestTextFilename("InstructionInfo", "MemorySizeInfo.txt");
		ArrayList<String> lines = FileUtils.readAllLines(filename);
		MemorySizeInfoTestCase[] result = new MemorySizeInfoTestCase[IcedConstants.MEMORY_SIZE_ENUM_COUNT];
		int lineNo = 0;
		for (String line : lines) {
			lineNo++;
			if (line.length() == 0 || line.charAt(0) == '#')
				continue;

			MemorySizeInfoTestCase tc;
			try {
				tc = parseLine(line, lineNo);
			}
			catch (Exception ex) {
				throw new UnsupportedOperationException(String.format("Invalid line %d (%s): %s", lineNo, filename, ex.getMessage()));
			}
			if (result[tc.memorySize] != null)
				throw new UnsupportedOperationException(String.format("Duplicate test, %s, line %d", filename, lineNo));
			result[tc.memorySize] = tc;
		}

		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < result.length; i++) {
			MemorySizeInfoTestCase tc = result[i];
			if (tc == null) {
				if (sb.length() > 0)
					sb.append(", ");
				sb.append(i);
			}
		}
		if (sb.length() != 0)
			throw new UnsupportedOperationException(String.format("Missing tests in %s: %s", filename, sb.toString()));
		return result;
	}

	private static MemorySizeInfoTestCase parseLine(String line, int lineNo) {
		int expValue = 6;
		if (MiscInstrInfoTestConstants.MEMORY_SIZE_ELEMS_PER_LINE != expValue)
			throw new UnsupportedOperationException();
		String[] elems = line.split(",", MiscInstrInfoTestConstants.MEMORY_SIZE_ELEMS_PER_LINE);
		if (elems.length != MiscInstrInfoTestConstants.MEMORY_SIZE_ELEMS_PER_LINE)
			throw new UnsupportedOperationException(String.format("Expected %d commas", MiscInstrInfoTestConstants.MEMORY_SIZE_ELEMS_PER_LINE - 1));

		MemorySizeInfoTestCase tc = new MemorySizeInfoTestCase();
		tc.lineNumber = lineNo;
		tc.memorySize = ToMemorySize.get(elems[0].trim());
		tc.size = NumberConverter.toInt32(elems[1].trim());
		tc.elementSize = NumberConverter.toInt32(elems[2].trim());
		tc.elementType = ToMemorySize.get(elems[3].trim());
		tc.elementCount = NumberConverter.toInt32(elems[4].trim());
		for (String value : elems[5].split(" ")) {
			if (value.equals(""))
				continue;
			Integer flags = InstrInfoDicts.toMemorySizeFlags.get(value);
			if (flags == null)
				throw new UnsupportedOperationException(String.format("Invalid flags value: %s", value));
			tc.flags |= flags;
		}
		return tc;
	}
}
